### PR TITLE
[llama4]: flatten dp_mesh to enable data parallel replicate

### PR DESCRIPTION
**Pull request by lkhphuc**
Oneliner fix to enable training with HSDP.

When DP Replicate is enabled, dp_mesh contains `(dp_shard_cp, dp_replicate)`, and `dp_mesh.get_group()` will throw an error due to multiple axes.
So I flatten it the same way it is done in `build_device_mesh()` function, with the same name just to be explicit).

Note: I should be able to use `world_mesh['dp_cp']` here, as it's already flatten when world_mesh is built. However inside the function when it is executed, this flatten mesh is not available for some reason.

### CODE REVIEW

File: torchtitan/experiments/llama4/infra/parallelize_llama.py [hunk 1]
```diff
@@ -124,7 +124,7 @@ def parallelize_llama(
         def _sync_tokens_per_expert(module, *_):
             assert isinstance(module, MoE)
             torch.distributed.all_reduce(
-                module.tokens_per_expert, group=dp_mesh.get_group()
+                module.tokens_per_expert, group=dp_mesh._flatten('dp_cp').get_group()
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/infra/parallelize_llama.py]: Calling `_flatten` every time the `all_reduce` happens sounds strange -- @fegin is the overhead repeated every time?
Anyways we may want to create a dedicated 1d mesh `dp_cp` when setting up device mesh.

Comment (fegin) [reply torchtitan/experiments/llama4/infra/parallelize_llama.py]: There are some CPU overheads, but no NCCL/GPU overhead. But it is better to do this upfront as @tianyu-l mentioned. The performance is not a major concern, deadlock is.

Comment (lkhphuc) [reply torchtitan/experiments/llama4/infra/parallelize_llama.py]: >  Anyways we may want to create a dedicated 1d mesh dp_cp when setting up device mesh.

We already have that device mesh.
https://github.com/pytorch/torchtitan/blob/315191e95a5ca9f1005301a7e025cd6e97519ff0/torchtitan/distributed/parallel_dims.py#L85-L103

If I try to use `world_mesh["dp_cp"].get_group()` inside the hook, it will complain that dim does not exist.
I found that if it's assigned to a variable (`dp_mesh` here), it can be used inside the hook.
I also delete the redundant logic to get `dp_mesh_dim_names` as well.

Comment (fegin) [reply torchtitan/experiments/llama4/infra/parallelize_llama.py]: Uh, not supporting this slicing is one design choice  some specific reasons way back ago. We should also improve DeviceMesh. cc., @wconstab

File: torchtitan/experiments/llama4/infra/parallelize_llama.py [hunk 2]
```diff
@@ -80,11 +80,7 @@ def parallelize_llama(
     if (
         parallel_dims.dp_shard_enabled or parallel_dims.cp_enabled
     ):  # apply FSDP or HSDP, potentially with Context Parallel
-        if parallel_dims.dp_replicate_enabled:
-            dp_mesh_dim_names = ("dp_replicate", "dp_shard_cp")
-        else:
-            dp_mesh_dim_names = ("dp_shard_cp",)
-        dp_mesh = world_mesh[tuple(dp_mesh_dim_names)]
+        dp_mesh = world_mesh["dp_cp"]
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/infra/parallelize_llama.py]: This doesn't look right. `fully_shard` for HSDP expects a 2D (sub)mesh, the first dim for `dp_replicate`, the second dim for `dp_shard_cp`

Comment (lkhphuc) [reply torchtitan/experiments/llama4/infra/parallelize_llama.py]: Ah sorry I missed that.
Now I reverted to old behaviour for `fully_shard`, only assigning the `world_mesh['dp_cp']` to use with the backward hook only.

I hope this is correct now.

File: torchtitan/experiments/llama4/infra/parallelize_llama.py [hunk 3]
```diff
@@ -119,12 +118,13 @@ def parallelize_llama(
 
     # for MoE auxiliary-loss-free load balancing
     if dp_mesh is not None:
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/infra/parallelize_llama.py]: we should condition on reason rather than result
```suggestion
    if parallel_dims.dp_replicate_enabled or parallel_dims.dp_shard_enabled or parallel_dims.cp_enabled:
```
we can have a cached property in `parallel_dims` for this long testing as it's also used at https://github.com/pytorch/torchtitan/blob/main/torchtitan/train.py#L446, but I don't have a good name... (maybe `dp_cp_enabled`?) we can do it later.

File: torchtitan/experiments/llama4/infra/parallelize_llama.py [hunk 4]
```diff
@@ -85,7 +85,6 @@ def parallelize_llama(
         else:
             dp_mesh_dim_names = ("dp_shard_cp",)
         dp_mesh = world_mesh[tuple(dp_mesh_dim_names)]
-
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/infra/parallelize_llama.py]: maybe restore?

**Review by tianyu-l — COMMENTED** [review]
LGTM, please fix final comment

**Comment by lkhphuc** [comment]
I created a new property `parallel_dims.dp_cp_enabled` and used it in train.py as well as parallelize llama 4.

A good candidate to name dp_cp could be `token_parallel` and `token_parallel_mesh`. I found myself converting dp_cp_mesh in my head to token mesh all the time since it is the mesh where any of my tokens of shape `BxS` can be sharded, e.g along batch dim B for data parallel and along seq dim S for Context parallel. WDYT?

Edit: Also add a new line before `def _sync_tokens_per_expert(...)` to pass linter.

**Review by tianyu-l — APPROVED** [review]
Looks good.

> A good candidate to name dp_cp could be token_parallel and token_parallel_mesh.

That's reasonable. I think `dp_cp_enabled` is not too bad at this moment. Let's land it and see how people receive it. If it causes confusion, we can consider renaming.
